File: 559. Maximum Depth of N-ary Tree/solution.py
"""
# Definition for a Node.
class Node(object):
  def __init__(self, val=None, children=None):
    self.val = val
    self.children = children
"""

class Solution(object):
  def maxDepth(self, root):
    """
    :type root: Node
    :rtype: int
    """
    def dfs(current, depth):
      if current == None:
        return depth - 1
      maxDepth = depth
      for child in current.children:
        currentDepth = dfs(child, depth + 1)
        if currentDepth > maxDepth:
          maxDepth = currentDepth
      return maxDepth

    return dfs(root, 1)

File: 559. Maximum Depth of N-ary Tree/solution.cpp
/*
// Definition for a Node.
class Node {
public:
  int val;
  vector<Node*> children;

  Node() {}

  Node(int _val) {
    val = _val;
  }

  Node(int _val, vector<Node*> _children) {
    val = _val;
    children = _children;
  }
};
*/

class Solution {
  int dfs(Node* current, int depth){
    if (current == NULL){
      return depth - 1;
    }
    int maxDepth = depth;
    for (int index = 0; index < current->children.size(); index++){
      Node* child = current->children[index];
      int currentDepth = dfs(child, depth + 1);
      if (currentDepth > maxDepth){
        maxDepth = currentDepth;
      }
    }
    return maxDepth;
  }
public:
  int maxDepth(Node* root) {
    return dfs(root, 1);
  }
};